### DISCUSSION TITLE
reader.TryGetField(string, out field) not getting first column

### DISCUSSION THREAD

**Discussion by snowfire20** [JoshClose/CsvHelper#1216] [Q&A]
I have a very simple csv file that i would like to parse by column names (header record) using TryGetValue(string, out field).this works correctly for the second and third columns, but not the first.
It

 appears that the NamedIndexCache does not have the first column, although Named indexes does. (although you can't get it by key). 
this finds a value: reader.Context.NamedIndexes.Where(n=>n.Key.Equals(FileColumns.SettingCode, StringComparison.InvariantCultureIgnoreCase))
this does not find a value. reader.Context.NamedIndexes.Where(n=>n.Key == FileColumns.SettingCode)

I think there may be control characters being read in with the first column header.

using csvhelper 12.1.1, and a csv generated by excel 2017
below is the code, and attached are screenshots of my tests.
[settingsimport.csv.txt](https://github.com/JoshClose/CsvHelper/files/2757229/settingsimport.csv.txt)
![equals](https://user-images.githubusercontent.com/6811262/51143763-473fea00-181d-11e9-9f29-2ae1e48761d7.PNG)

![keynotfound](https://user-images.githubusercontent.com/6811262/51142955-4908ae00-181b-11e9-9830-32accd02af08.PNG)

![namedindex](https://user-images.githubusercontent.com/6811262/51142387-d1864f00-1819-11e9-9903-d13462066b28.PNG)

```
	private static class FileColumns
		{
			public const string SettingCode = "SettingCode";
			public const string Value = "Value";
			public const string Type = "Type";

		}
		string[] Columns = { FileColumns.SettingCode, FileColumns.Value, FileColumns.Type };
		private List<Results>  ProcessUploadedFile(StreamReader streamReader, string delimeter)
		{
			var results = new List<Results>();
			var codes = SettingCode.FindAll<SettingCode>();
			using (IReader csvReader = new CsvReader(streamReader, new CsvHelper.Configuration.Configuration { Delimiter = delimeter }))
			{
				while (csvReader.Read())
				{
					if (csvReader.Context.Row == 1)//header row
					{
						csvReader.ReadHeader();

						if (csvReader.Context.HeaderRecord.Length != Columns.Length)
						{
							results.Add(new Results(){Message="The csv file does not contain 6 columns as expected"});
							return results;
						}

						if (!Columns.ContainsAny(csvReader.Context.HeaderRecord))
						{
							results.Add(new Results() { Message = "The csv file columns are invalid or not as expected" });
							return results;
						}
					
						continue;
					}

					results.Add(validateRow(csvReader, codes, results));
				}
			}
			return results;

		}
private Results validateRow(IReader reader, List<SettingCode> codes, List<Results> results)
		{
			var result = new Results() { Row = reader.Context.Row };
			
			string field;

			if (reader.TryGetField(FileColumns.SettingCode, out field) && !string.IsNullOrWhiteSpace(field))
			{
				result.SettingCode = field.Trim();
			}

			if (reader.TryGetField(FileColumns.Value, out field) && !string.IsNullOrWhiteSpace(field))
			{
				result.Value = field.Trim();
			}
			if (reader.TryGetField(FileColumns.Type, out field) && !string.IsNullOrWhiteSpace(field))
			{
				result.Type = field.Trim();
			}
	return result;
}
```

**Comment by JoshClose** [comment]
Can you break this down into a small self contained example that fails? Something similar to this.

```cs
void Main()
{
	var s = new StringBuilder();
	s.AppendLine("SettingCode,Value,Type");
	s.AppendLine("joke,me,jerk");
	s.AppendLine("fudge,me,fork");
	using (var reader = new StringReader(s.ToString()))
	using (var csv = new CsvReader(reader))
	{
		while (csv.Read())
		{
			if (csv.Context.Row == 1)
			{
				csv.ReadHeader();
				continue;
			}
			
			string field;
			csv.TryGetField("SettingCode", out field);
			field.Dump();
			
			csv.TryGetField("Value", out field);
			field.Dump();
			
			csv.TryGetField("Type", out field);
			field.Dump();
		}
	}
}
```

Leave out all the extra stuff and narrow it down to the failure.

**Comment by snowfire20** [comment]
I will try to get to this, bogged down with work right now

**Comment by joelenelson** [comment]
I am having this issue as well.  The .csv file in question is created by a Linux machine and therefore uses LF as the only new line character (as opposed to also using CR as with Windows).  When parsing this file with LF characters in a Windows environment, the TryGetField function fails to retrieve any values in the first field for the entire file.

If I add a garbage/filler column as the first column, everything works fine for my implementation.  Also, if I create the file in Windows with the same information, everything works fine.  For me, it's only when the LF character is the only new line character and the app is running in Windows that TryGetField fails.

Edit:  I believe it might be an encoding issue.  Because if I encode the file in UTF-8, instead of UTF-8-BOM, the file is parsed fine regardless of which character is used for designating a new line.

**Comment by AltruCoder** [comment]
@joelenelson Would it work to specifically tell it to detect the encoding from byte order marks?
```csharp
using (var reader = new StreamReader("path\\to\\file.csv", detectEncodingFromByteOrderMarks: true))
using (var csv = new CsvReader(reader))
{    
    var records = csv.GetRecords<Foo>();
}
```

**Comment by joelenelson** [comment]
> @joelenelson Would it work to specifically tell it to detect the encoding from byte order marks?
> 
> ```cs
> using (var reader = new StreamReader("path\\to\\file.csv", detectEncodingFromByteOrderMarks: true))
> using (var csv = new CsvReader(reader))
> {    
>     var records = csv.GetRecords<Foo>();
> }
> ```

That would work great.  Except that my implementation uses StringReader instead of StreamReader.  No such encoding overload exists for StringReader.

**Comment by joelenelson** [comment]
I changed my implementation to use a stream instead of a string, and to detect the encoding as suggested, and the encoding problem is gone.  Thank you.
